Enable export of code metrics in Markdown and HTML

aravind666 opened the pull request:
As part of request for [new feature ](https://github.com/philips-software/cerberus/issues/1) Hound JCMD-DIFF  now allows users to generate metrics report in Markdown and HTML format. 

As part of implementing this change, I also identified that going forward it will be difficult if we do not bring in some good AOP container in place. As a first step I did some research and figured out in all way Google Guice is a best fit for us and I have already created another branch called ressurect-IOC-AOP feel free to start contributing this branch . 

@dsvellal , @kryptt, @simaos  please review

aravind666 committed:
Enabled exporting of codemetrics diff in markdown and HTML format

aravind666 committed:
Code cleanup and fixed few PMD bugs and killed fue mutants

aravind666 committed:
Code cleanup and fixed few PMD bugs and killed fue mutants

dsvellal commented on src/main/java/com/philips/swcoe/cerberus/hounds/JavaCodeMetricsWithDiff.java:
```diff
@@ -61,15 +61,23 @@ public Integer call() throws Exception {
+
+    private void validateReportFormat(String metricFormat) {
+        if("horizontal".equalsIgnoreCase(metricFormat)) {
+            throw new CommandLine.ParameterException(spec.commandLine(), INVALID_REPORT_FORMAT);
```
Should we log an exception message, which says: Invalid report format found, defaulting to CSV_REPORT (or something like that), and then continue instead of breaking the flow, when: 
a) a non-valid reporting format is specified, or
b) when no reporting format is specified?

aravind666 replied:
We should not, because defaulting to certain behaviour allows cultivation of bad usage and puts pressure on us in future when we want to do some changes. Keeping less default behaviour helps us to keep the design evolutionary by nature. 

For eg.. I default to CSV here, and tomm some thing else comes in for horizontal report, we get into soup where we cannot do this change with out causing the break in usage. for eg ... if some one has scripted a shell script with these parameters and they think that we are anyway defaulting to CSV and post doing a new release we should also tell them to update explicitly and they might even revolt back saying why can't we continue defaulting to CSV which is waste. 

Embedding default behaviours in the software, is a kind of an antipattern which needs to be avoided as much as possible to prevent future wastes, and keep things as evolutionary as possible.

dsvellal commented on src/test/java/com/philips/swcoe/cerberus/hounds/JavaCodeMetricsWithDiffTest.java:
```diff
@@ -192,14 +198,28 @@ public void shouldThrowErrorForBadReportFormat() throws Exception {
+        int exitCode = new CommandLine(javaCodeMetricsWithDiff).execute(new String[]{
+                "--current", currentPath,
+                "--previous", previousPath,
+                "--format", "md",
```
Did you not add html reporter as well? May be worth adding a test for that as well?

aravind666 replied:
Yess, thanks for identifying this,  it is there , but in the service level, but you are write, I missed out adding the HTML in the message . I found some more issues in this suite of tests. Just corrected it now check

trieshard commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/AbstractCodeMetricsWriterService.java:
```diff
@@ -12,55 +15,67 @@
+    protected Builder getVerticalMarkdownPrinter() {
+        Builder tableBuilder = new Builder();
+        tableBuilder.withAlignments(Table.ALIGN_CENTER, Table.ALIGN_CENTER, Table.ALIGN_CENTER, Table.ALIGN_CENTER, Table.ALIGN_CENTER, Table.ALIGN_CENTER);
+        tableBuilder.addRow("FILE", "CLASS", "TYPE", "METRIC", "NEW_VALUE", "OLD_VALUE");
```
If field list is same for getVerticalPRinterWithDelimiter and here, consider extracting the list.

aravind666 replied:
Thanks for identifying this, I have now abstracted it out and used that method, please check the latest commit

trieshard commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/results/CodeMetricsClassResult.java:
```diff
@@ -7,53 +7,51 @@
+    private CodeMetricsDiffResult subClassesQty;
+    private CodeMetricsDiffResult lambdasQty;
+    private CodeMetricsDiffResult uniqueWordsQty;
+    private CodeMetricsDiffResult numberOfMethods;
```
Consistent naming here (is either numberOf or Qty) will likely be better for readers of this code to know if these are all the same (counts of some metric) or that there is something special or different between numberOf and Qty.

aravind666 replied:
Thanks for identifying, this Anti pattern, You are correct and I made it consistant accross the metrics package, by ensuring that every Quantifying variable in the metrics is named as variableCount . Please check ..

trieshard commented on src/main/java/com/philips/swcoe/cerberus/hounds/JavaCodeMetricsWithDiff.java:
```diff
@@ -61,15 +61,23 @@ public Integer call() throws Exception {
+            this.writeToUI(new CodeMetricsVerticalWriterService(classConfig, methodConfig, reportFormat.toUpperCase()).generateMetricsReport(metricsResult));
         } else {
-            this.writeToUI(new CodeMetricsHorizontalWriterService(classConfig, methodConfig, delimiter).generateMetricsReport(metricsResult));
+            this.writeToUI(new CodeMetricsHorizontalWriterService(classConfig, methodConfig, reportFormat.toUpperCase()).generateMetricsReport(metricsResult));
```
Updates (like this commit) are becoming expensive because each involves editing all near-duplicate pairs of logic for method and class level metrics. This cost will only increase as updates increase and the cost to refactor later will also increase. Consider refactoring for arbitrary grouping of metrics (class, method, whatever) and express only the small variations in the code instead. They seem ~90% similar.

aravind666 replied:
You are correct @simaos , This should be last, I have abstracted out Metrics generator and Metrics Reporter, and now that abstraction process is complete. From now on any new format of report we can just add a new reporter.  The new reporter is a factory and has behaviours to generate report in the intended format .

trieshard commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/AbstractCodeMetricsWriterService.java:
```diff
@@ -12,55 +15,67 @@
 import java.util.stream.Stream;
 
-public class AbstractCodeMetricsWriterService  {
+public class AbstractCodeMetricsWriterService {
```
Why not mark this abstract because it is only used as base for horizontal and vertical writers. Not intended for instantiation directly if I understand the context correctly.

trieshard replied:
Also consider removing the naming noise. Name it as what it is CodeMetricsWriter. Does the Abstract or Service add any additional meaning? Especially once you mark the class abstract. Possibly you can also lose the "Code" because the context is clear for this application - Cerberus is a code metrics application.

aravind666 replied:
@simaos , thanks for identifying this, Long long ago , this was an abstract class :)  and now it is not and I have missed to rename it . Now it is renamed . W.R.T removing Code from the context, I thought through this but in lots of our conversations we use the word metrics which sounds like a common noun, but I do agree that the term "codemetrics" too is a common noun but in this context it made sense for me to keep code as a prefix.

trieshard replied:
@aravind666 thanks. Do you intend only renaming? Why not actually add the abstract modifier to the type? Is there some code that I did not see that needs to make an instance of this class? Looks like only horizontal and verical writers have instances. This is abstract base for both of those.

aravind666 replied:
You are right @simaos thanks for identifying this, there is no place where we instantiate, but it is ideally not so good to call some thing an abstract when we don't have any thing which is abstract with in it. I think that was the reason I removed abstract modifier for this class. It is also correct that this class is the parent class and no one else should instantiate it and I should some how enforce  that only child classes could instantiate this. So I have made the constructor of this class as protected . Now things should look clean

trieshard commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/CodeMetricsHorizontalWriterService.java:
```diff
@@ -15,9 +15,9 @@
 public class CodeMetricsHorizontalWriterService extends AbstractCodeMetricsWriterService  {
```
See comment earlier on name of AbstractCodeMetricsWriterService. If you clean this naming noise then here you can rename also this class to HorizonalMetricsWriter. So we have a nice flow for the reader to understand, i.e., "This is a kind of metrics writer for my code metrics application", "What kind of metrics writer?", "A horizontal metrics writer". Simple and aligned with how you would explain the relationship verbally of you were talking someone through the code to explain.

aravind666 replied:
Yes, it is now corrected

dsvellal commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/AbstractCodeMetricsWriterService.java:
```diff
@@ -12,55 +15,67 @@
-        return new CSVPrinter(sw, csvFormat);
+    protected CSVPrinter getVerticalPrinterWithDelimiter(String format, StringWriter sw) throws IOException {
+        try {
+            char delimiter =  ProgramConstants.DelimeterForReport.valueOf(format.toUpperCase()).getDelimeter();
```
A spelling mistake for Delimeter? Should it be Delimiter? 

-- On a side-note, should we add a spell-check in our github actions?

aravind666 replied:
Corrected thanks

dsvellal commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/AbstractCodeMetricsWriterService.java:
```diff
@@ -12,55 +15,67 @@
-    protected boolean doesItMatterToDisplay(CodeMetricsResult codeMetricsResult, List<String> methodMetricsToDisplay) {
-        Boolean doesItMatter = codeMetricsResult.getOldValue() != codeMetricsResult.getNewValue();
-        return shouldWeDisplay(methodMetricsToDisplay, codeMetricsResult) && doesItMatter;
+    protected boolean doesItMatterToDisplay(CodeMetricsDiffResult codeMetricsDiffResult, List<String> methodMetricsToDisplay) {
```
When I was trying to understand why we'd have two methods: doesItMatterToDisplay, and shouldWeDIsplay(), I had to read through logic of both methods to understand the logic within them. I found an opportunity to split the responsibility a bit better. Something like this -> 

```java
boolean isMetricConfiguredToBeDisplayed = listOfMetricsToDisplay.contains(codeMetricsDiffResult.getMetricName()) || listOfMetricsToDisplay.isEmpty();
boolean isNewMetricValueDifferentFromOld = codeMetricsResult.getOldValue() != codeMetricsResult.getNewValue();
boolean shouldWeDisplay = isMetricConfiguredToBeDisplayed && isNewMetricValueDifferentFromOld;
```

The above refactoring captures what we are doing in each of those methods and provide context to the user reading the code without going through the details of the method.

trieshard replied:
I had same cognitive load to follow the logic. Especially ambiguous is that listOfMetricsToDisplay.isEmpty really means show all metrics. Alternatives include populating this list if user does not specify (i.e., show all) or have wrapper class for metrics to display instead of raw list, then ask it, i.e., "metrics.shouldDisplay(string name)" rather than a different class containing this display logic.

trieshard replied:
I do not think performance is an issue for this application (yet) but you do not need to check both. If either fails, the metric will not be shown. You can make a composite boolean and use Java short-circuit processing order for && to deal with only running the second check if needed. For example: return metrics.shouldDisplay(metric) && mertic.hasChanged()). Three benefits: lower cognitive load and logic for what should display is in its natural home (the configured metrics list) and logic for metric has changed is in its natural home (the metric). Right now you need to pull the data out of the metric result object to do the conditional. "Feature envy" maybe. Depends if you need metric result class to be pure data for another reason.

aravind666 replied:
Thanks @dsvellal You are correct these names are really not so correct and creates a confusion . I have now changed it please check .

aravind666 replied:
@simaos thanks for your comments, you are correct I have made it as OR since what ever turns true all conditions get evaluated, and if either gets evaluated to true, I am rendering the metrics .  You are right I am using metric result as pure data object.

dsvellal commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/CodeMetricsHorizontalWriterService.java:
```diff
@@ -15,9 +15,9 @@
-    public CodeMetricsHorizontalWriterService(List<String> classConfig, List<String> methodConfig, char delimiter) throws IOException {
-        super(classConfig, methodConfig, delimiter);
-        csvPrinter = getCsvPrinterHorizontal(this.delimiter, reportWriter);
+    public CodeMetricsHorizontalWriterService(List<String> classConfig, List<String> methodConfig, String reportFormat) throws IOException {
```
Should we consider passing report format as an enum? This way: 
1. We do not allow any arbitrary value to be passed to the report format
2. If an arbitrary value is passed via a command-line - we can catch it at the earliest state, default it to a format, and then continue everywhere as an enum

Would it be worth leveraging the ReportGenerator enum that's created already?

aravind666 replied:
Thats a nice idea, but Generator is too different from writer. Writer is common for all formats of data and generator is not common and we have enum since we have different way to generate report for each format

dsvellal commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/CodeMetricsVerticalWriterService.java:
```diff
@@ -14,59 +14,56 @@
-    public CodeMetricsVerticalWriterService(List<String> classConfig, List<String> methodConfig, char delimiter) throws IOException {
-        super(classConfig, methodConfig, delimiter);
-        csvPrinter = getCsvPrinterVertical(this.delimiter, reportWriter);
+    public CodeMetricsVerticalWriterService(List<String> classConfig, List<String> methodConfig, String reportFormat) throws IOException {
```
Same comment as above for the reportFormat variable -> enum v/s String, with defaulting enum at the earliest to report in a default format if an arbitrary value is passed as a command-line input.

Would it be worth leveraging the ReportGenerator enum that's created already?

dsvellal commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/CodeMetricsVerticalWriterService.java:
```diff
@@ -14,59 +14,56 @@
-        csvPrinter = getCsvPrinterVertical(this.delimiter, reportWriter);
+    public CodeMetricsVerticalWriterService(List<String> classConfig, List<String> methodConfig, String reportFormat) throws IOException {
+        super(classConfig, methodConfig, reportFormat);
+        csvPrinter = getVerticalPrinterWithDelimiter(reportFormat, reportWriter);
```
This is no longer a "CSV Printer" right? - it can be any delimiter separated. Should we change the name of the variable to match the intention?

aravind666 replied:
we do have printers readly available if user wants it . This printer is consumed by generator

dsvellal commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/CodeMetricsVerticalWriterService.java:
```diff
@@ -14,59 +14,56 @@
+        getStreamMetricsMethods(codeMetricsResult.getClass().getMethods()).forEach(methodInResult -> {
+            Try.of(() -> (CodeMetricsDiffResult) methodInResult.invoke(codeMetricsResult))
+                                .andThen(codeMetricsResultForConstruct -> Try.run(() -> writeMetrics(codeMetricsResultForConstruct, config))
+                                .onFailure(exception -> log.trace(exception.getStackTrace())));
```
May be worth considering onFailure: log-execption & default value of metrics?

aravind666 replied:
Thanks for identifying this, I am tracing because it helps us to support people if they share the trace of program flow. Since I am handling exception in monadic way, It becomes hard for us to support if some one shares just an exception .

dsvellal commented on src/main/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/CodeMetricsVerticalWriterService.java:
```diff
@@ -14,59 +14,56 @@
+    }
+
+    private void writeMetrics(CodeMetricsDiffResult codeMetricsDiffResult, List<String> config) throws IOException {
+        if(null == csvPrinter) {
```
Tomorrow, if we'd want another type of write (eg: HTML) would it be good to abstract out the writers into a Factory pattern based on the format specified by command-line, and inject the writer and have only one method? 

Would it be worth leveraging the ReportGenerator enum that's already created?

aravind666 replied:
Nope Writer is different, and Generator is different, To write CSV or PSV we use CSV writer , during the process of instantiation of this object by birth it will possess the required writer. 
using the writer that it possess, we write the metrics. Post writing the metrics. We generate a report. Based on the writer we generate the report . 

In our context think like this, writer is the pen, and generator is the one which ensures things written or embosed on paper.  Depending on the pen that you use generator chooses the paper and embose the content on to it.

dsvellal commented on src/test/java/com/philips/swcoe/cerberus/cerebellum/codemetrics/java/CodeMetricsHorizontalWriterServiceTest.java:
```diff
@@ -38,19 +38,19 @@ public void setupJavaCodeMetricsWithDiff() throws IOException {
     @Test
     public void shouldWriteClassMetricsandMethodMetricsInHorizontalFormatInCSV() throws Exception {
-        getCodeClassResult(',');
+        getCodeClassResult("CSV");
```
May be create an enum for this?

dsvellal reviewed (commented):
Please find my review comments. Would be happy to hear your thoughts.

aravind666 committed:
Addressed CodeReview Comments

	1. Renamed AbstractCodeMetricsWriterService.java to CodeMetricsWriterService.java since it is no more an abstract class
	2. Removed unused file ReportWriters.java

aravind666 committed:
Addressed code review comments of Simao

	1. Removed naming Noise in the metrics module
	2. We now have a standard inplace and got consistancy across the
module that any Quantifying variables will be named as variablenameCount
	3. Removed a duplicate initialiazing of list of strings

aravind666 committed:
Addressed Codereview comments from Simao and Datta

	1. Made the constructor of the base class
CodeMetricsWriterService.java as protected so that I can enforce that
only child class instantiation

	2. Missed out HTML report format in the validation message and
also in hound tests it is now added

aravind666 committed:
Fixed code review comments from Datta

aravind666 committed:
Updated readme and documentation for new formats MD and HTML

aravind666 committed:
Updated styleguide of checkstyle to enforce coding standards

ghost commented:
Nice changes recently - thanks. Curious why some classes end in "Service". Why not only the name? MetricsWriter vs MetricsWriterService?

aravind666 commented:
Good catch @simaos , I need to think about changing this convention, because the initial idea was to have everything inside Cerbellum package as services . If you see for factories we specify it as factory.
